sqlx-core/src/postgres/listen.rs

```diff
@@ -0,0 +1,216 @@
+        let _ = conn.send(cmd.as_str()).await?;
+        Ok(PgListener::new(conn))
+    }
+
+    /// Fetch a new connection from the pool and register it as a listener on the specified channels.
+    pub async fn listen_all(
+        &self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<PgPoolConnection>> {
+        let mut conn = self.acquire().await?;
+        let cmd = build_listen_all_query(channels);
+        let _ = conn.send(cmd.as_str()).await?;
+        Ok(PgListener::new(conn))
+    }
+}
+
+impl PgPoolConnection {
+    /// Register this connection as a listener on the specified channel.
+    ///
+    /// If an error is returned here, the connection will be dropped.
+    pub async fn listen(mut self, channel: impl AsRef<str>) -> Result<PgListener<Self>> {
+        let cmd = build_listen_all_query(&[channel]);
+        let _ = self.send(cmd.as_str()).await?;
+        Ok(PgListener::new(self))
+    }
+
+    /// Register this connection as a listener on all of the specified channels.
+    ///
+    /// If an error is returned here, the connection will be dropped.
+    pub async fn listen_all(
+        mut self,
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<Self>> {
+        let cmd = build_listen_all_query(channels);
+        let _ = self.send(cmd.as_str()).await?;
+        Ok(PgListener::new(self))
+    }
+}
+
+/// A stream of async database notifications.
+///
+/// Notifications will always correspond to the channel(s) specified this object is created.
+pub struct PgListener<C>(C);
+
+impl<C> PgListener<C> {
+    /// Construct a new instance.
+    pub(self) fn new(conn: C) -> Self {
+        Self(conn)
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: DerefMut<Target = PgConnection>,
+{
+    /// Get the next async notification from the database.
+    pub async fn next(&mut self) -> Result<NotifyMessage> {
+        loop {
+            match (&mut self.0).receive().await? {
+                Some(Message::NotificationResponse(notification)) => return Ok(notification.into()),
+                Some(msg) => {
+                    return Err(protocol_err!(
+                        "unexpected message received from database {:?}",
+                        msg
+                    )
+                    .into())
+                }
+                None => continue,
+            }
+        }
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: Connection,
+{
+    /// Close this listener stream and its underlying connection.
+    pub async fn close(self) -> BoxFuture<'static, Result<()>> {
+        self.0.close()
+    }
+}
+
+impl<C> std::ops::Deref for PgListener<C> {
+    type Target = C;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<C> std::ops::DerefMut for PgListener<C> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<C: Connection<Database = Postgres>> crate::Executor for PgListener<C> {
```

Comment:
Not something you should fix in your PR but I'm tired of how much this trait is delegated. We need a macro. Haha.